// @tensorflow-models/depth-estimation/dist/ar_portrait_depth/calculators/segment_foreground.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.segmentForeground = void 0;
var tf = require("@tensorflow/tfjs-core");
/**
 * A calculator for changing the background of an image to black based on
 * results of segmentation returned by bodySegmentation package.
 * @param image Input image.
 * @param segmentation Segmentation mask of foreground/background.
 *
 * @returns Masked image.
 */
function segmentForeground(image, segmentation) {
    if (tf.getBackend() === 'webgl') {
        // Same as implementation in the else case but in one custom shader on GPU.
        return segmentForegroundWebGL(image, segmentation);
    }
    // tf.tidy is unnecessary since this function is called within a tf.tidy
    // context.
    var _a = image.shape, height = _a[0], width = _a[1], channels = _a[2];
    // Extract red channel which stores probability
    var probability = tf.slice(segmentation, 0, [height, width, 1]);
    // Round to make a binary mask that can be multiplied with image.
    var binaryMask = tf.round(probability);
    // Make the same shape as input image.
    var imageBinaryMask = tf.tile(binaryMask, [1, 1, channels]);
    return tf.mul(image, imageBinaryMask);
}
exports.segmentForeground = segmentForeground;
function segmentForegroundWebGL(image, segmentation) {
    var program = {
        variableNames: ['image', 'segmentation'],
        outputShape: image.shape,
        userCode: "\n  void main() {\n      ivec3 coords = getOutputCoords();\n      int height = coords[0];\n      int width = coords[1];\n      int channel = coords[2];\n      float value = getImage(height, width, channel);\n      float foregroundProbability = getSegmentation(height, width, 0);\n      setOutput(foregroundProbability >= 0.5 ? value : 0.0);\n    }\n"
    };
    var webglBackend = tf.backend();
    var outputTensorInfo = webglBackend.compileAndRun(program, [image, segmentation]);
    return tf.engine().makeTensorFromDataId(outputTensorInfo.dataId, outputTensorInfo.shape, outputTensorInfo.dtype);
}
